core/src/main/java/pl/allegro/tech/embeddedelasticsearch/EmbeddedElastic.java

```diff
@@ -33,6 +33,8 @@
 
     private ElasticServer elasticServer;
     private ElasticRestClient elasticRestClient;
+    private boolean started = false;
```

Comment:
Instead of two fields and DCL in start/stop methods, you can use AtomicBoolean. It will be much simpler

Reply:
I'm rusty on my java concurrency but I've used a synchronized block for a reason - to invalidate thread cache and flush other CPU cores / other thre that might have pending changes for that object. `AtomicBoolean` will not solve that AFAIK

Reply:
What do you mean by "pending changes for that object"?

Reply:
Now I see that both my proposition and your solution share same problem: when first thread reaches line which changes started state to true, second thread will finish method and think that Elasticsearch is ready, but it's not. 

I think it can be solved with something like this:

```
public EmbeddedElastic start() throws IOException, InterruptedException {
    synchronized (lifecycleLock) {
        if (!started) {
            installElastic();
            startElastic();
            createRestClient();
            createTemplates();
            createIndices();
            started = true; // started must be volatile
        }
    }
}

public void stop() {
    if (elasticServer != null) {
        synchronized (lifecycleLock) {
            if (started) {
                started = false;
                elasticServer.stop();
            }
        }
    }
}
```

BTW I think we might we can resign from lifecycleLock and simply make start/stop methods synchronized.

Reply:
> BTW I think we might we can resign from lifecycleLock and simply make start/stop methods synchronized.

yes!